feat: add git hook command

AdeAttwood opened the pull request:
feat: add git hook command

Summary:

This command is compatible with the V1 git hook so everything we have still
works. This implements a custom command runner and writer that will run shell
commands in parallel and display the output in a nice way. This is a full
replacement for the previous implementation and packages.

Test Plan:

This has been tested manually and the tests have been run. Right now this is
still a bit work in progress but it's a good start.

Ref: #77

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/Practically/Conventional-Tools/pull/86).
* #90
* #89
* __->__ #86
* #85

GeorgeCadwallader commented on tests/lib/runner/command.spec.ts:
```diff
@@ -0,0 +1,57 @@
+      '# The command title\necho "Goodbye, World!"',
+    );
+
+    it('sets the command body to be the hole script', () => {
```
**typo:** "hole" > "whole"

```suggestion
    it('sets the command body to be the whole script', () => {
```

GeorgeCadwallader commented on tests/lib/runner/command.spec.ts:
```diff
@@ -0,0 +1,57 @@
+    });
+  });
+
+  describe('with an already create command passed in', () => {
```
**typo:** I'm assuming this is meant to be:

```suggestion
  describe('with an already created command passed in', () => {
```

AdeAttwood committed:
feat: implement sapling source control provider

Summary:

You can now use conventional tools in a sapling repo (maybe).

Test Plan:

This has not really been tested yet, but the code is there. This is very much a
WIP, the users will be using this at their own risk.

AdeAttwood committed:
refactor: add root command to get the root of the current repository

Summary:

Right now in our git hooks we are using `git rev-parse --show-toplevel` to get
the root of the repository. This is not working within a sapling repo.

This implements the root command so we can use that instead. It as a new `root`
method to the `SourceControlProvider` interface. This will then be implemented
by any provider we want to.

Test Plan:

No real way to test this other than running the command and checking the
output. As long as the hooks commands will be working for us this will be fine
for now.

AdeAttwood committed:
feat: add git hook command

Summary:

This command is compatible with the V1 git hook so everything we have still
works. This implements a custom command runner and writer that will run shell
commands in parallel and display the output in a nice way. This is a full
replacement for the previous implementation and packages.

Test Plan:

This has been tested manually and the tests have been run. Right now this is
still a bit work in progress but it's a good start.

Ref: #77
